### PR TITLE
feat: Make OAuth2 Microsoft Provider configureable

### LINKED ISSUE #926: OAuth2 Tenant Configuration (Microsoft)
## 🚀 Feature

Hey there,

I discovered a problem where I tried to run an internal business application in Appwrite.
This application uses OAuth2 as authentication method against Microsoft Azure-AD.
The problem here is that internal business applications usually only allow the authentication of users who have already been created in the organization in the microsoft azure ad.
By default, the tenant "common" is hard-coded in the URL in Appwrite.
As a result, the application in Microsoft Azure AD has to be configured as a multi-tenant application, which of course is not always wanted, since access should only take place exclusively for company employees.
Therefore I would like to have an option to configure the tenant type so that you have the option to choose between "common" and "organization".

## Have you spent some time to check if this issue has been raised before?

Yes, the only solution currently is to adjust the tenant type manually in Microsoft.php, or to make the application multi-tenant capable.

### Have you read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/master/CODE_OF_CONDUCT.md)?

Yes.

### PR DESCRIPTION
**Pull request by BobcatProgrammer**
## What does this PR do?

This PR enables configuration of the TenantId for the Microsoft OAuth2 provider as documented [here](https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols#endpoints). This enables control over what types of accounts can authenticate via Microsoft.

To facilitate this the PR also changes how custom OAuth2 providers can be configured. Details can be found in the [docs](https://github.com/appwrite/appwrite/blob/master/docs/tutorials/add-oauth2-provider.md)

### Open Points in DRAFT

- [x] Make the JS part generic
- [x] Migrate Apple Provider to new structure
- [x] Update [Docs](https://github.com/appwrite/appwrite/blob/master/docs/tutorials/add-oauth2-provider.md)

## Test Plan

- [x] Test Microsoft Provider with `common`
- [x] Test Microsoft Provider with `organizations`
- [x] Test Microsoft Provider with `consumers`
- [x] Test Microsoft Provider with `{my tenant id}`
- [x] Test Apple provider Form works with generic JS

## Related PRs and Issues

Fixes #926 
Prepares for #1668 

### Have you read the [Contributing Guidelines on issues](https://github.com/appwrite/appwrite/blob/master/CONTRIBUTING.md)?

yes of course, even multiple times 😃

### CODE REVIEW

File: app/views/console/users/index.phtml [hunk 1]
```diff
@@ -475,9 +477,12 @@ $smtpEnabled = $this->getParam('smtpEnabled', false);
...
                                     <?php else: ?>
-                                        <label for="oauth2<?php echo $this->escape(ucfirst($provider)); ?>Appid">Bundle ID <span class="tooltip" data-tooltip="Attribute internal display name"><i class="icon-info-circled"></i></span></label>
-                                        <input name="appId" id="oauth2<?php echo $this->escape(ucfirst($provider)); ?>Appid" type="text" autocomplete="off" data-ls-bind="{{console-project.provider<?php echo $this->escape(ucfirst($provider)); ?>Appid}}" placeholder="com.company.appname" />
-                                        <input name="secret" data-forms-oauth-apple id="oauth2<?php echo $this->escape(ucfirst($provider)); ?>Secret" type="hidden" autocomplete="off" data-ls-bind="{{console-project.provider<?php echo $this->escape(ucfirst($provider)); ?>Secret}}" />
+                                        <?php
+                                            $form_view = new View(__DIR__.'/oauth/'.$this->escape($form));
```
Comment (eldadfux) [inline app/views/console/users/index.phtml:481]:
```suggestion
                                            $form = new View(__DIR__.'/oauth/'.$this->escape($form));
```

We use camelCase and try to stick to shorter names when possible.

File: app/views/console/users/index.phtml [hunk 2]
```diff
@@ -475,9 +477,12 @@ $smtpEnabled = $this->getParam('smtpEnabled', false);
...
+                                            echo $form_view
```
Comment (eldadfux) [inline app/views/console/users/index.phtml:482]:
```suggestion
                                            echo $form;
```

File: app/views/console/users/index.phtml [hunk 3]
```diff
@@ -475,9 +477,12 @@ $smtpEnabled = $this->getParam('smtpEnabled', false);
...
+                                                ->setParam("provider",$provider)
```
Comment (eldadfux) [inline app/views/console/users/index.phtml:483]:
```suggestion
                                                ->setParam("provider", $provider)
```

**Review by eldadfux — CHANGES_REQUESTED** [review]
Added a few small suggestions. Would be great to get a screenshot of the new form.

**Comment by BobcatProgrammer** [comment]
image
Screenshot of Microsoft OAuth Provider Form

**Review by eldadfux — APPROVED** [review]

**Comment by eldadfux** [comment]
Please try to sync with master, tests might be failing because of other build issues the core team had.

**Comment by BobcatProgrammer** [comment]
> Please try to sync with master, tests might be failing because of other build issues the core team had.

done! can someone retrigger the checks please?

File: src/Appwrite/Auth/OAuth2/Microsoft.php [hunk 4]
```diff
@@ -141,4 +141,39 @@ protected function getUser(string $accessToken): array
...
+     */
+    protected function decodeJson():array{
+        
+        try {
+            $secret = \json_decode($this->appSecret, true);
+        } catch (\Throwable $th) {
+            throw new Exception('Invalid secret');
+        }
+        return $secret;
+    }
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Microsoft.php:168]:
```suggestion
    protected function decodeJson():array
    {    
        try {
            $secret = \json_decode($this->appSecret, true);
        } catch (\Throwable $th) {
            throw new Exception('Invalid secret');
        }
        return $secret;
    }
```

File: src/Appwrite/Auth/OAuth2/Microsoft.php [hunk 5]
```diff
@@ -141,4 +141,39 @@ protected function getUser(string $accessToken): array
...
+
+    /**
+     * Extracts the Tenant Id from the JSON stored in appSecret. Defaults to 'common' as a fallback
+     * @return string
+     */
+    protected function getTenantId():string
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Microsoft.php:174]:
```suggestion
    protected function getTenantId(): string
```

File: src/Appwrite/Auth/OAuth2/Microsoft.php [hunk 6]
```diff
@@ -141,4 +141,39 @@ protected function getUser(string $accessToken): array
 
         return $this->user;
     }
+
+    /**
+     * Extracts the Client Secret from the JSON stored in appSecret
+     * @return string
+     */
+    protected function getClientSecret():string
```
Comment (christyjacob4) [inline src/Appwrite/Auth/OAuth2/Microsoft.php:149]:
```suggestion
    protected function getClientSecret(): string
```

**Review by christyjacob4 — CHANGES_REQUESTED** [review]
Amazing work on the PR! Just some minor PSR compliance and we're good to merge 👍

**Review by eldadfux — APPROVED** [review]

**Comment by eldadfux** [comment]
Amazing work @BobcatProgrammer, thank you so much for making this contribution to the project!

**Pull request merged**